diff --color -ruN sa/sqlite3.c sb/sqlite3.c
--- sa/sqlite3.c	2026-06-12 18:02:49.092973364 +0300
+++ sb/sqlite3.c	2026-06-12 18:02:49.108457426 +0300
@@ -74409,6 +74409,7 @@
   sqlite3_mutex_leave(pMainMtx);
   return removed;
 #else
+  UNUSED_PARAMETER(pBt);
   return 1;
 #endif
 }
@@ -82876,6 +82877,7 @@
 ** sqlite_schema table. Otherwise SQLITE_OK.
 */
 SQLITE_PRIVATE int sqlite3BtreeSchemaLocked(Btree *p){
+  UNUSED_PARAMETER(p);
   int rc;
   assert( sqlite3_mutex_held(p->db->mutex) );
   sqlite3BtreeEnter(p);
