diff --color -ruN sa/include/sbsms.h sb/include/sbsms.h
--- sa/include/sbsms.h	2026-06-12 18:02:49.180418531 +0300
+++ sb/include/sbsms.h	2026-06-12 18:02:49.185561816 +0300
@@ -49,7 +49,11 @@
 class SBSMSInterface /* not final */ {
  public:
   virtual ~SBSMSInterface() {}
-  virtual long samples(audio *buf, long n) { return 0; }
+  virtual long samples(audio *buf, long n) { 
+    (void)buf;
+    (void)n;
+    return 0;
+  }
   virtual float getStretch(float t)=0;
   virtual float getMeanStretch(float t0, float t1)=0;
   virtual float getPitch(float t)=0;
@@ -79,11 +83,22 @@
  public:
   virtual ~SBSMSRenderer() {}
   virtual void startFrame() {}
-  virtual void startTime(int c, const TimeType &time, int n) {}
-  virtual void render(int c, SBSMSTrack *t) {}
-  virtual void endTime(int c) {}
+  virtual void startTime(int c, const TimeType &time, int n) {
+    (void)c;
+    (void)time;
+    (void)n;
+  }
+  virtual void render(int c, SBSMSTrack *track) {
+    (void)c;
+    (void)track;
+  }
+  virtual void endTime(int c) {
+    (void)c;
+  }
   virtual void endFrame() {}
-  virtual void end(const SampleCountType &samples) {}
+  virtual void end(const SampleCountType &samples) {
+    (void)samples;
+  }
 };
 
 enum SBSMSError {
diff --color -ruN sa/src/fft.h sb/src/fft.h
--- sa/src/fft.h	2026-06-12 18:02:49.177644035 +0300
+++ sb/src/fft.h	2026-06-12 18:02:49.182920903 +0300
@@ -6,7 +6,7 @@
 #include <string.h>
 #include "utils.h"
 
-#if defined(ENABLE_SSE) && !defined(APPLE_PPC)
+#if defined(ENABLE_SSE) && !defined(APPLE_PPC) && !defined(_M_ARM64)
 #include "sse.h"
 #endif
 
@@ -322,7 +322,7 @@
   }
 };
 
-#if defined(ENABLE_SSE) && !defined(APPLE_PPC)
+#if defined(ENABLE_SSE) && !defined(APPLE_PPC) && !defined(_M_ARM64)
 
 template <int N, int dir> 
 class SSETwiddles {
