--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -97,7 +97,7 @@
 endif()
 
 target_link_libraries(mnxdom PUBLIC nlohmann_json::nlohmann_json)
-target_link_libraries(mnxdom PUBLIC nlohmann_json_schema_validator)
+target_link_libraries(mnxdom PUBLIC nlohmann_json_schema_validator::validator)
 # nlohmann/json behavior config
 target_compile_definitions(mnxdom PUBLIC JSON_DISABLE_ENUM_SERIALIZATION=1)
 # Export fetched mnx_w3c locations to dependents
