Audacity commit 5ba6072bbb
Subject: Workaround for bug #2264

diff --git a/nyquist/xlisp/xlprin.c b/nyquist/xlisp/xlprin.c
index 260363f..3eecf88 100644
--- a/nyquist/xlisp/xlprin.c
+++ b/nyquist/xlisp/xlprin.c
@@ -209,7 +209,7 @@ LOCAL void putqstring(LVAL fptr, LVAL str)
     for (p = getstring(str); (ch = *p) != '\0'; ++p)
 
         /* check for a control character */
-        if (ch < 040 || ch == '\\' || ch > 0176 || ch == '"') {
+        if (ch < 040 || ch == '\\' || ch > 0176 /* || ch == '"' */) {
             xlputc(fptr,'\\');
             switch (ch) {
             case '\011':
