Audacity commit f5ac7bb104
Subject: Typo in alpassvc.alg

diff --git a/nyquist/tran/alpassvc.alg b/nyquist/tran/alpassvc.alg
index 9578154..395b54a 100644
--- a/nyquist/tran/alpassvc.alg
+++ b/nyquist/tran/alpassvc.alg
@@ -29,7 +29,7 @@
             register sample_type *yptr;\n")
 (INNER-LOOP "/* compute where to read y, we want y to be delay_snd samples
              * after delay_ptr, where we write the new sample. First, 
-             * conver from seconds to samples. Note: don't use actual sound_type
+             * convert from seconds to samples. Note: don't use actual sound_type
              * names in comments! The translator isn't smart enough.
              */
             delaysamp = (sample_type) delaysnd * delay_scale_factor;
