Audacity commit c4e391cfc4
Subject: Move debug defines after including standard headers

diff --git a/nyquist/nyqsrc/convolve.c b/nyquist/nyqsrc/convolve.c
index e513bb2..b52534d 100644
--- a/nyquist/nyqsrc/convolve.c
+++ b/nyquist/nyqsrc/convolve.c
@@ -61,12 +61,6 @@
  * Length of output is length of x input + length of h
  */
 
-// You can turn on debugging output with: #define D if (1)
-#define D if (0)
-
-#define MAX_IR_LEN 4000000 /* maximum impulse response length */
-#define MAX_LOG_FFT_SIZE 16 /* maximum fft size for convolution */
-//#define MAX_LOG_FFT_SIZE 4 /* maximum fft size for convolution */
 #define _USE_MATH_DEFINES 1 /* for Visual C++ to get M_LN2 */
 #include <assert.h>
 #include <math.h>
@@ -84,8 +78,17 @@
 #include "fftext.h"
 #include "convolve.h"
 
-void convolve_free(snd_susp_type a_susp);
+#ifdef D
+#undef D // We would like to use D in this file
+#endif   // D
 
+// You can turn on debugging output with: #define D if (1)
+#define D if (0)
+#define MAX_IR_LEN 4000000 /* maximum impulse response length */
+#define MAX_LOG_FFT_SIZE 16 /* maximum fft size for convolution */
+//#define MAX_LOG_FFT_SIZE 4 /* maximum fft size for convolution */
+
+void convolve_free(snd_susp_type a_susp);
 
 typedef struct convolve_susp_struct {
     snd_susp_node susp;
